feat(flattenConcurrentlyAtMost): add new extra

xtianjohns opened the pull request:
Add flattenSequentiallyAtMost(n) extra.

### Description
Add new extra operator file and unit tests. The operator extra started as a copy of `flattenConcurrently` wrapped in a factory function. From there, a few renames and some combined functionality from `flattenSequentially` to implement the inner buffer/queue named `_seq`.

There are some distinctions between both, obviously. The completion method (`_c`) needs to be able to determine if there are any currently connected streams, or any streams in the buffer. If the answer to both questions is no, then we send a completion event downstream.

Additionally, there may be a case when the last inner stream completes, there's no items left in the queue, and the meta stream has completed. This also means we should send a completion event downstream.

To track these fields, the operators gets some new properties, `_l` which tracks how many streams the operator is listening to, `_seq` which represents the buffer and `_d` which indicates whether the meta stream has completed.

Resolves staltz/xstream#161.

### Motivation and Context
flattenSequentiallyAtMost is designed to provide consumer-configurable concurrency to flattening operations. Two flattening extras exist which allow consumers to flatten a meta stream with maximum concurrency, or with no concurrency. This new operator supports a concurrency limit, representing the maximum amount of _additional_ streams to connect to during flattening.

### How Has This Been Tested?
I've added 2 new tests which verify that if an inner stream errors that we send an error downstream, and that we don't greedily connect to more than `n` streams at a time.

I've also duplicated both test suites from `flattenSequentially` and `flattenConcurrently`, and substituted those extras with `flattenConcurrentlyAtMost(1)` and `flattenConcurrentlyAtMost(Infinity)` respectively. This may be overkill, but was a powerful tool during development to know I wasn't violating semantics by adding the configurable buffer.

## Checklist:
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

staltz commented on src/extra/flattenConcurrentlyAtMost.ts:
```diff
@@ -0,0 +1,132 @@
+      if (!u) return;
+      u._c();
+    }
+    if (this._l === 0 && seq.length > 0) {
```
Why do we need this `this._l === 0` check? Suppose `n === 3` and `this._l === 3`, once this `less()` is called, now `this._l === 2` and `2 < 3` so we should try to get stuff from the seq.

xtianjohns replied:
Good catch, that's a bug. We don't want to exhaust the list before draining the queue, I'll fix.

xtianjohns commented:
Thanks @staltz! I dunno what happened with the message (I got the branch name right), I've rebased the change and I've also grep'd the diff for the name and made sure I'm using the right language.

I've updated the limit to reflect the max for number of streams to listen to, so now `flattenSequentially` and `flattenConcurrentlyAtMost(1)` are equivalent.

staltz commented:
Thanks @xtianjohns! I noticed a few other issues, mostly with docs/tests. Comments inline

staltz commented on src/extra/flattenConcurrentlyAtMost.ts:
```diff
@@ -0,0 +1,132 @@
+ *    \       ----1----2---3--|
+ *    --a--b----c----|
+ *     flattenConcurrentlyAtMost(1)
+ * -----a--b----c------2---3--|
```
The output here *could* be correct, as well as it could have been different, e.g.

```
-----a--b----c---------1----2---3--|
```

In case there was no third-party subscriber for the 1-2-3 inner stream. I think this diagram should be the one I just showed because flattenConcurrentlyAtMost(1) is supposed to be like flattenSequentially, and if you see the diagram for that one, it's a-b-c-1-2-3.

xtianjohns replied:
Right, good thinking. Makes sense to leave the diagrams as simple as possible, and not assuming other subscribers seems like a good approached. Fixed!

staltz commented on src/extra/flattenConcurrentlyAtMost.ts:
```diff
@@ -0,0 +1,132 @@
+ *      \   -----1----2----3--|
+ *       ---a--b-----c--|
+ *     flattenConcurrentlyAtMost(2)
+ * ---------a--b-1---c2--i-3---jh--|
```
Like the previous diagram, I think this one should be 

```
--+---+---+-|
   \   \   \
    \   \   ---fgh----i-----jh--|
     \   -----1----2----3--|
      ---a--b-----c--|
    flattenConcurrentlyAtMost(2)
---------a--b-1---c2--i-3------fgh----i-----jh--|
```

xtianjohns replied:
Fixed!

staltz commented on tests/extra/flattenConcurrentlyAtMost.ts:
```diff
@@ -0,0 +1,387 @@
+describe('flattenConcurrentlyAtMost (extra)', () => {
+
+  describe('with n less than Infinity', () => {
+    it('should process only n + 1 streams at a time', (done: any) => {
```
I think this should be `only n streams`

xtianjohns replied:
Agree! Fixed.

xtianjohns commented:
Yep, these make sense @staltz, will fix.

xtianjohns commented on tests/extra/flattenConcurrentlyAtMost.ts:
```diff
@@ -0,0 +1,387 @@
+        // ---00--01
+        // --------10------11
+        // -----------20----------21
+        const expected = ['00', '01', '10', '20', '11', '21'];
```
Good catch, I've added this exact case: that the operator subscribes to new streams from its buffer only when it needs new events.

xtianjohns commented:
Sorry for the wait, @staltz. You may be even _less_ likely to merge now since it's been so long, but hey - at least you can decline a complete PR. 😄

staltz commented:
@xtianjohns Thanks for the reminder! No reason to decline this PR, I probably just lost it in the middle of many other notifications. :)

staltz commented:
Released 11.9.0
